ToggleGroupControl: rewrite backdrop animation with framer motion shared layout animations

ciampo opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Rewrite `ToggleGroupControl`s backdrop animation using `framer-motion`'s [shared layout animations](https://www.framer.com/motion/layout-animations/#shared-layout-animations).

This PR supersedes https://github.com/WordPress/gutenberg/pull/49822, since in the meantime:
- `framer-motion` got updated to a more recent version (https://github.com/WordPress/gutenberg/pull/49822)
- `ToggleGroupControl` went through some major internal refactoring (https://github.com/WordPress/gutenberg/pull/45123 in particular)


## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

The current implementation for the backdrop animation is complex, difficult to read, imperative (based off arbitrary timeouts), and [buggy](https://github.com/WordPress/gutenberg/issues/37506).

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->


Main changes are:

- deleted the `ToggleGroupControlBackdrop`, and the imperative logic that came with it
- re-implemented the backdrop using `framer-motion`'s shared layout animations:
    - the main way this works is by using the same value for the `layoutId` attribute to tag the components that participate to the same shared layout animation
    - each `ToggleGroupControl` also defines its own "namespace" for the shared layout animations via the `LayoutGroup` component, in order to avoid conflicts between different instances of the component
- cleanup up some styles and removed some unused variables
- memoized some classnames
- tweaked the  `InspectorControl`'s `Slot` and `Fill` components to forward the `MotionContext`
- rewrote the internal controlled/uncontrolled logic, that previously relied on `usePrevious` + `useUpdateEffect`. The refactor meant:
  - using `useControlledValue` for the "as button" version of the component
  - refactoring from `reakit` to `ariakit` for the "as radio" version of the component (since it has support for controlled/uncontrolled logic)
  - adding some extra logic to distinguish between controlled/uncontrolled when an `undefined` value is passed and based on the previous value of the `value` prop (to maintain backwards-compat)

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

- In storybook, go over all the examples for the `ToggleGroupControl` component and make sure that the backdrop animates as expected
- In the post editor, select a paragraph block, and make sure that the font size picker (rendering a `ToggleGroupControl`) works and animates as expected
- In the site editor, open the global styles sidebar:
  - Select "Typography", then one of the list elements. Make sure that the font size picker (rendering a `ToggleGroupControl`) works and animates as expected
  - Select a heading. Make sure that the font size picker (rendering a `ToggleGroupControl`) works and animates as expected
### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->


https://github.com/WordPress/gutenberg/assets/1083581/69b6c703-0a78-4b63-8cfd-fd7020215957

ciampo committed:
Reuse previous implementation

ciampo committed:
Reset motion context in slot-fill

ciampo committed:
Memoize classes

ciampo committed:
Move layout group to root component

ciampo committed:
Use useLayoutEffect to apply internal state changes

ciampo committed:
CHANGELOG

ciampo committed:
Update snapshots

ciampo commented:
Next steps / ideas:
- Play around with the `layout` and `layoutRoot` prop (see [docs](https://www.framer.com/motion/layout-animations/###item-is-performing-unwanted-layout-animations-as-surrounding-content-changes))
- See if it's possible to somehow disable the initial animation (I tried the `initial` prop and it didn't seem to work)
- See if it's possible to reduce the layout jumps in the sidebar by tweaking those components

(cc @youknowriad @mirka @aaronrobertshaw @andrewserong @stokesman in case you either have capacity to experiment with this, or have other ideas)

ciampo committed:
Expose motion context from the components package

ciampo committed:
Refactor inspector controls slot/fill to accept an array of context provider/values under fillprops.forwardedContext

ciampo committed:
Add framer motion context to the forwarded contexts in inspector controls slot/fill

ciampo commented:
Update: I should have managed to solve the remaining issues.


Regarding the backdrop not animating correctly inside `InspectorControls`: inspired by the recent changes suggested by @jsnajdr (see #51264), I forwarded the `MotionContext` of the `Slot` via `fillProps` (see [commits diff](https://github.com/WordPress/gutenberg/pull/50278/files/b593ca4adad9af111ee31b410c41b91efb5eaa16..94c1b970572ea0130bc5ba88d8e6806009ed08dd)), instead of simply resetting it.

Regargind the backdrop not animating correctly when `isDeselectable={true}`: it looks like I had forgotten to refactor part of the component from `useUpdateEffect` to a custom implementation using `useLayoutEffect` (instead of the `useEffect` used internally by `useUpdateEffect`) — I pushed the change in [this commit](https://github.com/WordPress/gutenberg/pull/50278/commits/85a4150bc640d3bd645b25ebb80e53da3ae917cc).

The last standing issue now is around the "deselectable" version of `ToggleGroupControl`, where sometimes the backdrop doesn't animate — although I think this PR is ready for at least a first round of review to understand if:

- the technical approach is solid
- the animation feels good, or if there's anything to tweak

FInally, after the update to `useLayoutEffect`, I had to also update unit tests to use the controlled version of the component. That's definitely something that we'd need to look into before merging, as it is potentially a regression.

ciampo commented on packages/components/src/index.ts:
```diff
@@ -17,7 +17,11 @@ export {
+export {
+	__unstableMotion,
+	__unstableAnimatePresence,
+	__unstableMotionContext,
```
TODO: we may want to export the motion context differently? (ie. stable, or as private API)

jsnajdr replied:
Let's export all the providers (style, components, toolbar context, motion) consistently the same way, as private APIs.

ciampo replied:
I'm not sure we can remove existing exports from the package — even if they are prefixed with `__unstable`, removing them would cause breaking changes in plugin and themes using them (I ran [a quick search on wpdirectory](https://wpdirectory.net/search/01H2GD26S6CXSS8CWX4PWN771G) and there are unfortunately some matches).

@youknowriad do you have any suggestions on how we may want to approach this? Given how "de facto" stable are `__unstableMotion` and `__unstableAnimatePresence`, should we just export them without the `__unstable` alias too?

ciampo committed:
Remove hardcoded motion provider from fill component, update slot fill storybook example

ciampo committed:
CHANGELOG update

ciampo committed:
Update as button group to also use useLayoutEffect

ciampo committed:
Use controlled version of the component in unit tests

jasmussen reviewed (approved):
The user experience is great. Before:

![before](https://github.com/WordPress/gutenberg/assets/1204802/65337c08-40f1-4b22-8db5-440e90676b11)

Notice here that the backdrop behind the font size toggle I click actually animates in from _nothing_/from the left, when the initial control is unset.

After:

![after](https://github.com/WordPress/gutenberg/assets/1204802/dde788b4-21a7-4f9e-bd6c-c38a9c417c9e)

Notice here how when the control is initiallly unset, the backdrop just fades in, instead of also animating in. This is an improvement in this PR, nice.

The animation is also slightly slower, that's fine.

jsnajdr commented on packages/block-editor/src/components/inspector-controls/fill.js:
```diff
@@ -73,15 +70,19 @@ function ToolsPanelInspectorControl( { children, resetAllFilter, fillProps } ) {
 		};
 	}, [ resetAllFilter, registerResetAllFilter, deregisterResetAllFilter ] );
```
This effect needs to be rewritten, too. The `fillProps` value used to be `toolsPanelContext`, but now the context has moved into the `forwardedContext` array. The `register`/`deregister` fields are no longer there.

You need to move the registration _under_ the `ToolsPanelContext.Provider`. Into the `innerMarkup` value:
```js
function RegisterResetAll( { resetAllFilter, children } ) {
  const { register, deregister } = useContext( ToolsPanelContext );
  useEffect( () => {
    if ( resetAllFilter && register && deregister ) {
      register( resetAllFilter );
      return () => { deregister( resetAllFilter ); };
    }
  }, [ resetAllFilter, register, deregister ] );
  return children;
};

const innerMarkup = <RegisterResetAll resetAllFilter={ ... }>{ children }</RegisterResetAll>;
```

ciampo replied:
Good point, done in https://github.com/WordPress/gutenberg/pull/50278/commits/e6754b23a9198f525f536e2732e18c05eea47129

Also, cc'ing @aaronrobertshaw to make sure that the registration/deregistration still works as expected!

aaronrobertshaw replied:
I couldn't see any issues with ToolsPanelItem registration or deregistration in my testing. At a glance, the code changes make sense to me also.

jsnajdr commented on packages/components/src/toggle-group-control/stories/index.tsx:
```diff
@@ -135,3 +139,142 @@ Deselectable.args = {
+		<ExampleSlotFill.Fill>
+			{ ( fillProps: {
+				forwardedContext?: [
+					React.Context< any >[ 'Provider' ],
```
This type is too constraining, we also have `<StyleProvider document={ ... }>` which doesn't satisfy this.

`forwardedContext` is an array of entries like:
```ts
type Entry = [ Provider<Props>, Props ]
```
where `Provider` is a React component that accepts `Props` and also must accept `children`. Therefore:
```ts
type Provider<Props> = React.Component<PropsWithChildren<Props>>;
```
That's all what `forwardedContext` must satisfy, nothing more.

ciampo replied:
`React.Component<React.PropsWithChildren<any>>` didn't work for me — I was getting an error on the line below where I rendered `<Provider { ...props }>{ inner }</Provider>`:

`JSX element type 'Provider' does not have any construct or call signatures.ts(2604)`


Changing the type to `React.FunctionComponent<React.PropsWithChildren<any>>` makes TSC happy (https://github.com/WordPress/gutenberg/pull/50278/commits/2e93f44a3aa7f3cc5d90ac9f7ba33dbac5e3410c)

Also worth noting that these extra storybook examples will be removed before merging the PR (which makes this specific type not super important), although I appreciate the review comment — learning every day!

jsnajdr replied:
`React.ComponentType` is even better, allows all kinds of components. That's what I should originally have written, `React.Component` is a different thing 🙂 

One day the context forwarding will be typed, so the discussion is relevant beyond one storybook.

The type for the props is still wrong: it requires `value`, but for `StyleProvider` valid props are `{ document }`. Also it would be nice if the provider and props types were to match each other. Currently it's `any` everywhere, that's not real typing 🙂

ciampo replied:
Something like this perhaps? https://github.com/WordPress/gutenberg/pull/50278/commits/edb79a5500a0ea8608a2eaf66369383be2ccfaf9

jsnajdr replied:
Yes, this is better :+1:

ciampo committed:
Improve context provider storybook types

ciampo committed:
Fix register/deregister logic for toolspanel in inspector controls

ciampo committed:
extract useUpdateLayoutEffect to separate util hook

ciampo committed:
Better fillprops forwarded context type

ciampo committed:
Change docs to have controlled usage

aaronrobertshaw reviewed (commented):
Nice work here @ciampo, the updated animation is great ✨ 

Everything is testing as advertised for me. 

I couldn't break it in regards to the ToolsPanelItem registration/deregistration. The ToolsPanel stories function as before and the component's unit tests are passing.

The ToggleGroupControl animations were nice and smooth in all the contexts I tested. 

I've run out of time today to give a complete code review however I can do that later this week if nobody beats me to it in the meantime.

ciampo committed:
Reuse types

ciampo committed:
Add defaultValue prop

ciampo committed:
Refactor component to handle both controlled and uncontrolled updates

ciampo committed:
Remove unused `useUpdateLayoutEffect` hook

ciampo committed:
Update unit tess: check both controlled and uncontrolled, use initialvalue

ciampo commented:
> Finally, after the update to `useLayoutEffect`, I had to also update unit tests to use the controlled version of the component. That's definitely something that we'd need to look into before merging, as it is potentially a regression.

I've looked a bit more into this, and effectively it looks like my previous changes introduced a regression.

I've spent some time looking into it, and the way I managed to restore support for both the uncontrolled and controlled usage of the component introduces a few changes:

- in order to detect controlled / uncontrolled, I moved away from using `usePrevious` (and `useUpdateLayoutEffect` to propagate changes). Instead, the component is now using `useControlledValue` for the `AsButtonGroup` case, and a custom version of the same logic for the `AsRadioGroup` case (this is because we still need to call `useRadioState`, and therefore I couldn't use `useControlledValue`)
- in order to allow consumers of the component to set an initial state without using the component in "controlled" mode, I introduced a new `defaultValue` prop
- as a consequence of the above points, I also changed the implementation of the `ToggleGroupItem` component when rendered as a deselectable button: when the button gets un-pressed, the `onChange` function is called with `''` instead of `undefined`. This is necessary to follow the expectations of the `useControlledValue` hook, where an `undefined` value is associated with the uncontrolled usage of the component


A few considerations / questions:

@diegohaz : would you mind checking the way I've implemented controlled/uncontrolled support around `useRadioState` ?  If we migrated to ariakit, would there be any changes?

@mirka:
- I know that this proposed solution moves away from the logic that you had implemented with `usePrevious`. I tried to keep that logic around but I wasn't able to have the units tests passing otherwise, after refactoring to framer motion
- Do you think that the proposed API changes make sense? ie.
  - Introduction of the `defaultValue` prop, to be used to set the component's initial value (works both in controlled and uncontrolled mode)
  - Changed the expected value for "controlled but empty" from `undefined` to `''` (as described above). We could also use other values, such as `null`, what do you think?


If we're ok with these changes but you find the PR's scope too large, I'm also happy to split the "controlled / uncontrolled" refactor to a separate PR, to be merged before this one.

diegohaz commented:
> @diegohaz : would you mind checking the way I've implemented controlled/uncontrolled support around `useRadioState` ? If we migrated to ariakit, would there be any changes?

That should be the same as if it were `React.useState` instead of `Reakit.useRadioState`. So that looks OK.

It's one of the things that have improved a lot in Ariakit, which would look like this:

```js
const radio = useRadioStore( {
  defaultValue: initialValue,
  value: valueProp,
  setValue: onChangeProp,
} );
```

More info: https://ariakit.org/guide/component-stores

ciampo commented:
> That should be the same as if it were `React.useState` instead of `Reakit.useRadioState`. So that looks OK.

Thank you!

> It's one of the things that has been improved a lot in Ariakit, which would look like this:

Good to know! If you ever planned on refactoring this component to Ariakit soon, maybe we could do it after this refactor is complete?

diegohaz commented:
> Good to know! If you ever planned on refactoring this component to Ariakit soon, maybe we could do it after this refactor is complete?

Do you mean waiting for this PR to be merged before migrating `ToggleGroupControl` to Ariakit? Sure.

mirka commented on packages/components/src/toggle-group-control/toggle-group-control/as-button-group.tsx:
```diff
@@ -40,58 +38,38 @@ function UnforwardedToggleGroupControlAsButtonGroup(
-			groupContext.setState( value );
-		}
-	}, [ value ] );
+	const [ selectedValue, setSelectedValue ] = useControlledValue( {
```
Separate from the [discussion](https://github.com/WordPress/gutenberg/issues/47473#issuecomment-1498704142) of whether we want to lean into the `useControlledValue` logic, I'm a bit concerned about this as a breaking change at this stage. For example if somebody had a reset button like `<button onClick={ () => setValue( undefined ) }>Reset</button>`, it would no longer work.

Would it be possible to retain the original detection logic for this PR, or would it be too complicated?

ciampo replied:
Tentative approach added in https://github.com/WordPress/gutenberg/pull/50278/commits/8ea25492a68393b299b400b2ab5bd67b6469e3b3

I also updated unit tests and added a reset button for the controlled mode.

mirka commented on packages/components/src/toggle-group-control/toggle-group-control/component.tsx:
```diff
@@ -39,18 +41,25 @@ function UnconnectedToggleGroupControl(
 		() =>
 			cx(
-				styles.ToggleGroupControl( { isBlock, isDeselectable, size } ),
+				styles.toggleGroupControl( { isBlock, isDeselectable, size } ),
```
Nice catch 😄

mirka commented on packages/components/src/toggle-group-control/types.ts:
```diff
@@ -112,6 +112,10 @@ export type ToggleGroupControlProps = Pick<
 	 */
 	value?: ReactText;
+	/**
+	 * The selected value.
```
```suggestion
	 * The initial value to be used when in uncontrolled mode.
```

or something like that?

Also this needs to be added in the README.

ciampo replied:
Good point, done in https://github.com/WordPress/gutenberg/pull/50278/commits/9de1f76d34618093b764d98eed87235b33cac88a

mirka commented on packages/components/src/toggle-group-control/test/index.tsx:
```diff
@@ -213,26 +238,25 @@ describe( 'ToggleGroupControl', () => {
 				expect( mockOnChange ).toHaveBeenCalledTimes( 1 );
-				expect( mockOnChange ).toHaveBeenCalledWith( undefined );
-				expect(
+				expect( mockOnChange ).toHaveBeenLastCalledWith( '' );
```
I'm afraid this is also kind of a breaking change if somebody already has logic checking for `value === undefined`.

ciampo replied:
See https://github.com/WordPress/gutenberg/pull/50278#discussion_r1242444300

mirka commented on packages/components/src/toggle-group-control/toggle-group-control-option-base/component.tsx:
```diff
@@ -123,6 +146,19 @@ function ToggleGroupControlOptionBase(
 			</WithToolTip>
+			{ /* Animated backdrop using framer motion's shared layout animation */ }
+			{ isPressed ? (
+				<motion.div
```
Do we need to tweak the transition duration to match the original (`transitionDurationFast`)? It feels a lot slower than what we have in trunk.

ciampo replied:
Joen seems ok with this according to [his review above](https://github.com/WordPress/gutenberg/pull/50278#pullrequestreview-1471390529)

> The animation is also slightly slower, that's fine.

mirka reviewed (commented):
Awesome job on this, it's a lot cleaner and reliable! I like how it eliminated the awkward initial animation, too. It's working well in my testing.

The two main things I noticed was the breaking change for the `undefined` value, and the relatively slower transition duration compared to trunk.

ciampo committed:
Introduce logic to distinguish undefined as controlled or uncontrolled update

ciampo committed:
Bring back README example to be uncontrolled, but use `defaultValue`

ciampo committed:
Update `defaultValue` prop description, add to README

ciampo committed:
Stop using deprecated `ReactText` type, use `string | number` instead

ciampo commented:
Hey @mirka , I've managed to find a working solution that would allow us to keep allowing consumers of the component to receive and set `undefined` values on the component without breaking the controlled/uncontrolled logic.

The solution relies on interpreting an incoming `undefined` value as really undefined only if the component has never been used in controlled mode during its lifetime; otherwise, the `undefined` value will be interpreted as an empty value for controlled mode.

In case this solution works for you, I only see two open questions left before merging this PR:

- Are we ok exporting FramerMotion's context [this way](https://github.com/WordPress/gutenberg/pull/50278#discussion_r1223121220)?
- Should this PR be split into two smaller ones (internal refactor on how to handle controlled/uncontrolled, and then framer motion refactor), or are you ok to merge it as-is ? (before merging I will remove the extra Storybook examples, which account for ~80 lines of code added.

ciampo committed:
Add keyboard interaction with reset unit test

ciampo commented:
During my testing, I found a bug: when using arrow keys, resetting the value of `ToggleGroupControl` doesn't work


https://github.com/WordPress/gutenberg/assets/1083581/c952e908-3c59-4c45-a460-4ab02155d3d7



I did some debugging, and this seems to be related to some internal logic in `ariakit`. In particular, it seems like [this `useEffect`](https://github.com/ariakit/ariakit/blob/reakit%401.3.11/packages/reakit/src/Radio/Radio.ts#L134-L144) fires unexpectedly when clicking on the reset `Button`, likely because of the hook's `onChange` dependency changing.

The `onChange` function depends on the `setState` function that we set internally. Prior to this PR, the internal `setState` was the value returned by `useRadioState` and was a stable function. With the changes introduced by this and the need to support controlled mode, I'm not sure there's an easy way around it.

@diegohaz , would you be able to take a look and see if you can come up with any workaround? Or would it be just easier to migrate to `ariakit`, since it seems to support controlled/uncontrolled mode out of the box?

ciampo committed:
Add reset button to Storybook example

ciampo committed:
Refactor to ariakit

ciampo commented:
I tried to refactor the component from `reakit` to `ariakit`, which seems to fix the issue [highlighted above](https://github.com/WordPress/gutenberg/pull/50278#issuecomment-1609291587) (tests are passing locally for me)

.. which means that this PR is once more ready for a review

> In case this solution works for you, I only see two open questions left before merging this PR:
> 
> * Are we ok exporting FramerMotion's context [this way](https://github.com/WordPress/gutenberg/pull/50278#discussion_r1223121220)?
> * Should this PR be split into two smaller ones (internal refactor on how to handle controlled/uncontrolled, and then framer motion refactor), or are you ok to merge it as-is ? (before merging I will remove the extra Storybook examples, which account for ~80 lines of code added.

diegohaz commented:
> @diegohaz , would you be able to take a look and see if you can come up with any workaround? Or would it be just easier to migrate to `ariakit`, since it seems to support controlled/uncontrolled mode out of the box?

The fastest workaround I could think of was making the `onChange` function stable using a technique similar to the new [`useEffectEvent`](https://react.dev/reference/react/experimental_useEffectEvent) hook in React. Ariakit implements this as [`useEvent`](https://github.com/ariakit/ariakit/blob/224ad57c467eb22f0de47c9fe0d57112b18bfeff/packages/ariakit-react-core/src/utils/hooks.ts#L93-L115) and automatically applies it to event props.

diegohaz commented on packages/components/src/toggle-group-control/toggle-group-control-option-base/component.tsx:
```diff
@@ -111,18 +141,39 @@ function ToggleGroupControlOptionBase(
+					// `render` prop, the component is still expecting `input` props
 					<Radio
 						{ ...commonProps }
-						{
-							...otherContextProps /* these are only for Ariakit Radio */
-						}
-						as="button"
+						onFocus={ (
+							event: React.FocusEvent< HTMLInputElement >
+						) => {
+							// @ts-expect-error Even if we're passing `button` through the
+							// `render` prop, the component is still expecting `input` props
+							commonProps.onFocus?.( event );
+							if ( event.defaultPrevented ) return;
+							toggleGroupControlContext.setValue( value );
+						} }
+						render={ <button /> }
```
You can pass the `onFocus` prop directly to `<button />` to get the correct types.

ciampo replied:
Good point, done in https://github.com/WordPress/gutenberg/pull/50278/commits/2f3e1e77086647f3bfdaa5cbfef520886be792a5

ciampo committed:
Pass props to `button` instead of `radio`

mirka commented on packages/components/src/toggle-group-control/stories/index.tsx:
```diff
@@ -31,6 +35,7 @@ const meta: ComponentMeta< typeof ToggleGroupControl > = {
 		help: { control: { type: 'text' } },
 		onChange: { action: 'onChange' },
 		value: { control: { type: null } },
+		defaultValue: { control: { type: null } },
```
I was thinking about why we still need to add a `defaultValue` prop when we're supporting `undefined` values, and then realized we have another breaking change here 😭

On trunk, we used to be able to just give a "default value" to the `value` prop and use the component as uncontrolled:

```jsx
		<ToggleGroupControl
			{ ...props }
			onChange={ ( ...changeArgs ) => {
				onChange?.( ...changeArgs );
			} }
			value={ 'left' }
		/>
```

Could we maybe change the `hasEverBeenUsedInControlledMode` heuristics to see if the `value` has changed at least once? (Kind of seems like `usePrevious` might be inevitable 🤔)

ciampo replied:
Alright, I gave this approach a try in https://github.com/WordPress/gutenberg/pull/50278/commits/288eb0b14735ad54880a85ddd80f9fdf29c788a7

Let me know if it now looks good to you!

ciampo committed:
add support for uncontrolled without the `defaultValue` prop

ciampo committed:
Update snappshots

ciampo committed:
Restore README

ciampo committed:
Update CHANGELOG

ciampo committed:
Format

ciampo commented:
Hey @mirka , could you give this PR one last review? Hopefully [your latest feedback](https://github.com/WordPress/gutenberg/pull/50278#discussion_r1245234174) has been addressed correctly.

One thing that I'll note is that snapshots were failing because of extra `tabindex` props (see [commit](https://github.com/WordPress/gutenberg/pull/50278/commits/b8467d56dbcbb7e778ef6d80c76f1f7524ac6a93)) — not sure if it's a big deal though (it may be related to the refactor from reakit to ariakit).

ciampo committed:
Fix format (again)

jameskoster reviewed (approved):
Everything is working as advertised for me 👍

mirka commented on packages/components/src/toggle-group-control/toggle-group-control/component.tsx:
```diff
@@ -36,18 +38,24 @@ function UnconnectedToggleGroupControl(
+	const baseId = useInstanceId(
+		ToggleGroupControl,
+		'toggle-group-control'
+	).toString();
```
Why is the `toString()` necessary here?

mirka commented on packages/components/src/toggle-group-control/stories/index.story.tsx:
```diff
@@ -48,15 +52,20 @@ const Template: StoryFn< typeof ToggleGroupControl > = ( {
+			/>
+			<Button onClick={ () => setValue( undefined ) } variant="tertiary">
+				Reset
+			</Button>
```
The reset button is going to be removed before merging, right?

mirka reviewed (approved):
Thanks for all the improvements! Looks good and is testing well.

ciampo committed:
Remove unnecessary cast to string when using `useInstanceId`

ciampo committed:
Remove temporary Storybook changes

ciampo committed:
CHANGELOG

ciampo committed:
Fix indentation
